UIforETW/UIforETWDlg.cpp

```diff
@@ -51,13 +863,16 @@ void outputPrintf(_Printf_format_string_ const wchar_t* pFormat, ...)
 	va_end(args);
 }
 
-void CUIforETWDlg::vprintf(const wchar_t* pFormat, va_list args)
+void CUIforETWDlg::vprintf(PCWSTR pFormat, va_list args)
 {
 	wchar_t buffer[5000];
 	_vsnwprintf_s(buffer, _TRUNCATE, pFormat, args);
 
-	for (const wchar_t* pBuf = buffer; *pBuf; ++pBuf)
+	for (PCWSTR pBuf = buffer; *pBuf; ++pBuf)
 	{
+		ATLASSERT(pBuf < (buffer + ARRAYSIZE(buffer)));

```

Comment:
:-) I think it was Steve McConnell who wrote about the advantages of no-fail programming. He gave the example of asking various candidates to write a toupper function. Many of them incorporated error handling. But as soon as you do that - as soon as core functions like that can fail - then the necessary error handling becomes an exponential explosion of (untested) code paths. How much better it is to write a toupper that always works.

Of course, some functions can fail, and that may have to be handled. But having as few functions as possible that can fail, and knowing (or deciding, or mandating) what you can count on is very powerful. Learning how to compose large systems with as little error handling code as possible, while still handling all possible errors, is an art form.

Video games offer one example. They hook allocation functions at a low-level so that they will never return NULL or throw an exception. If an allocation fails then it is a fatal error and it is handled (deep inside the alloc function) with __debugbreak() or exit() or some-such. The error is handled, but at zero size and complexity cost in the client code.

And thus quicksand is avoided. This technique does not apply to all software, or at least not to all allocations in all software (if a user can specify an arbitrarily large image size then the application should handle a failure to allocate the image buffer) but it is powerful.